### PR TITLE
refactor(optimizer): divide logical optimizer into one for batch and one for streaming.

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Move logical optimization from `PlanRoot` to `LogicalOptimizer`.
- Use static rules instead of creating new ones for queries every time.
- Separate logical optimization into one for batch and one for streaming. 

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [x] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>
#8137

### CODE REVIEW

File: src/frontend/src/optimizer/logical_optimization.rs [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+use crate::optimizer::rule::{
+    AggDedupGroupKeyRule, AggProjectMergeRule, ApplyAggTransposeRule, ApplyFilterTransposeRule,
+    ApplyJoinTransposeRule, ApplyProjectTransposeRule, ApplyScanRule, ApplyShareEliminateRule,
+    ApplyToJoinRule, BoxedRule, DagToTreeRule, DistinctAggRule, FilterWithNowToJoinRule,
+    JoinCommuteRule, MaxOneRowEliminateRule, MergeMultiJoinRule, OverAggToTopNRule,
+    ProjectEliminateRule, ProjectJoinMergeRule, ProjectMergeRule, PullUpCorrelatedPredicateRule,
+    PushCalculationOfJoinRule, ReorderMultiJoinRule, RewriteLikeExprRule, TopNOnIndexRule,
+    TranslateApplyRule, TrivialProjectToValuesRule, UnionInputValuesMergeRule, UnionMergeRule,
+    UnionToDistinctRule,
+};
```
Comment (fuyufjh) [inline src/frontend/src/optimizer/logical_optimization.rs:36]:
```suggestion
use crate::optimizer::rule::*;
```

File: src/frontend/src/optimizer/logical_optimization.rs [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+    {
+        OptimizationStage {
+            stage_name: name.into(),
+            rules,
+            apply_order,
+        }
+    }
+}
+
+pub struct LogicalOptimization {}
```
Comment (fuyufjh) [inline src/frontend/src/optimizer/logical_optimization.rs:111]: How about `LogicalOptimizer`?

File: src/frontend/src/optimizer/logical_optimization.rs [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+        InputRefValidator.validate(plan.clone());
+
+        if ctx.is_explain_logical() {
+            ctx.store_logical(plan.explain_to_string().unwrap());
+        }
+
+        Ok(plan)
+    }
+
+    pub fn gen_optimized_logical_plan_for_batch(mut plan: PlanRef) -> Result<PlanRef> {
```
Comment (fuyufjh) [inline src/frontend/src/optimizer/logical_optimization.rs:419]: Hmmm. So it comes to another problem - lots of duplication code between the `for_batch` and `for_stream`. I am not sure whether it's better than before.

File: src/frontend/Cargo.toml [hunk 4]
```diff
@@ -31,6 +31,7 @@ fixedbitset = "0.4.1"
 futures = { version = "0.3", default-features = false, features = ["alloc"] }
 futures-async-stream = "0.2"
 itertools = "0.10"
+lazy_static = "1.4.0"
```
Comment (fuyufjh) [inline src/frontend/Cargo.toml:34]:
```suggestion
lazy_static = "1"
```

**Review by fuyufjh — COMMENTED** [review]
PATL @st1page

File: src/frontend/src/optimizer/logical_optimization.rs (hunk 3 above)
Comment (jon-chuang) [reply src/frontend/src/optimizer/logical_optimization.rs:410]: Somewhat agree. I think it's easier to understand in a single block and branching on batch or stream. It makes it clearer which optimizations are shared, and which apply only to batch or stream. 

But this assumes that the general ordering of optimization passes are shared (so far, it's been true, but I can't tell if this PR also changes that)

File: src/frontend/planner_test/tests/testdata/predicate_pushdown.yaml [hunk 5]
```diff
@@ -300,8 +302,10 @@
     create table t1(v1 timestamp with time zone, v2 int);
     select * from t1 where v1 > now() + '30 min' and v2 > 5;
   optimized_logical_plan: |
-    LogicalFilter { predicate: (t1.v1 > (Now + '00:30:00':Interval)) }
-    └─LogicalScan { table: t1, columns: [t1.v1, t1.v2], predicate: (t1.v2 > 5:Int32) }
+    LogicalJoin { type: LeftSemi, on: (t1.v1 > $expr1), output: all }
```
Comment (jon-chuang) [inline src/frontend/planner_test/tests/testdata/predicate_pushdown.yaml:305]: hmm, it's weird that the stream optimization is applied to optimized_logical_plan...?

File: src/frontend/planner_test/tests/testdata/agg.yaml [hunk 6]
```diff
@@ -701,18 +701,15 @@
     create table t(a int, b int, c int);
     select a, count(distinct b) as distinct_b_num, sum(distinct c) filter(where c < 100) as distinct_c_sum from t group by a;
   optimized_logical_plan: |
-    LogicalAgg { group_key: [t.a], aggs: [count(t.b) filter((flag = 0:Int64)), sum(t.c) filter((count filter((t.c < 100:Int32)) > 0:Int64) AND (flag = 1:Int64))] }
-    └─LogicalAgg { group_key: [t.a, t.b, t.c, flag], aggs: [count filter((t.c < 100:Int32))] }
-      └─LogicalExpand { column_subsets: [[t.a, t.b], [t.a, t.c]] }
-        └─LogicalScan { table: t, columns: [t.a, t.b, t.c] }
```
Comment (jon-chuang) [inline src/frontend/planner_test/tests/testdata/agg.yaml:707]: Hmm, is this only because we do 2-phase-agg for stream?

**Review by st1page — APPROVED** [review]
LGTM, We used to think the optimization for streaming and batch have a common phase and it save our test and coding for a while. But currently, we find that the difference between the two optimizations could be flexible. And after this pr change, we can plug in the different optimization for the two execution more flexibly.

> Hmmm. So it comes to another problem - lots of duplication code between the for_batch and for_stream. I am not sure whether it's better than before.

I think to reuse the duplication code, we should extract the common and meaningful functions such as `fn predicate_push_down(mut plan) ->Result<PlanRef>`(in fact this piece is duplicated even in one specific execution's optimizer) or `fn simple_push_down`. The emphasis of the "meaningful" here is because we do not think the batch and streaming shared a common optimization logical in nature. 

In summary, I think the changes are just like that we use the Combination to reuse the code instead of Inheritance here.

File: src/frontend/planner_test/tests/testdata/predicate_pushdown.yaml (hunk 5 above)
Comment (chenzl25) [reply src/frontend/planner_test/tests/testdata/predicate_pushdown.yaml:305]: Sorry, I have changed the `optimized_logical_plan` default meaning from the batch to the streaming before. I will change it back to the batch logical optimization.

File: src/frontend/planner_test/tests/testdata/agg.yaml (hunk 6 above)
Comment (chenzl25) [reply src/frontend/planner_test/tests/testdata/agg.yaml:707]: Ditto.

File: src/frontend/src/optimizer/logical_optimization.rs (hunk 3 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/logical_optimization.rs:410]: > Somewhat agree. I think it's easier to understand in a single block and branching on batch or stream. It makes it clearer which optimizations are shared, and which apply respectively only to batch or stream.
> 
> But this assumes that the general ordering of optimization passes are shared (so far, it's been true, but I can't tell if this PR also changes that)

In the beginning, it sounds good to only have a single block and branching on batch or stream, however, It contains more and more branches gradually. At least for now, I can give about 5 optimizations that needed to be applied to the batch and stream separately. In addition, I hope this separation can help developers to think of whether their optimization can only be applied to the batch or the streaming, or both.

File: src/frontend/planner_test/tests/testdata/predicate_pushdown.yaml (hunk 5 above)
Comment (st1page) [reply src/frontend/planner_test/tests/testdata/predicate_pushdown.yaml]: can we remove them and just maintain the `batch_plan` and `stream_plan`? or support `optimized_logical_plan_for_batch` and `optimized_logical_plan_for_stream`
